src/analysis/retail/shaman/elemental/modules/features/SpenderWindow.tsx

```diff
@@ -0,0 +1,308 @@
+import ThresholdPerformancePercentage, {
+  LTEThreshold,
+  GTEThreshold,
+} from './shared/ThresholdPerformancePercentage';
+import SPELLS from 'common/SPELLS';
+import TALENTS from 'common/TALENTS/shaman';
+import { Talent } from 'common/TALENTS/types';
+import { formatDuration } from 'common/format';
+import { SpellIcon, SpellLink, TooltipElement } from 'interface';
+import { explanationAndDataSubsection } from 'interface/guide/components/ExplanationRow';
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Events, { CastEvent } from 'parser/core/Events';
+import Enemies from 'parser/shared/modules/Enemies';
+import { ON_CAST_BUFF_REMOVAL_GRACE_MS } from '../../constants';
+
+interface ActiveSpenderWindow {
+  timestamp: number;
+  motePresent: boolean;
+  flameshockPresent: boolean;
+}
+
+interface FinishedSpenderWindow extends ActiveSpenderWindow {
+  elshocksPresent: boolean;
+  sopUse: CastEvent;
+}
+
+const SOP_CONSUME_SPELLS = [
+  SPELLS.LIGHTNING_BOLT,
+  TALENTS.CHAIN_LIGHTNING_TALENT,
+  SPELLS.FLAME_SHOCK,
+  TALENTS.FROST_SHOCK_TALENT,
+  TALENTS.LAVA_BURST_TALENT,
+];
+
+const GOOD_SOP_CONSUMERS = [SPELLS.LIGHTNING_BOLT.id, TALENTS.CHAIN_LIGHTNING_TALENT.id];
+
+const PERFECT_WINDOWS_THRESHOLD: GTEThreshold = {
+  type: 'gte',
+  perfect: 0.9,
+  good: 0.8,
+  ok: 0.7,
+};
+const WRONG_SOP_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.1,
+  ok: 0.15,
+};
+const MISSING_ELSHOCKS_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+const MISSING_MOTE_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+const MISSING_FLAMESHOCK_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+
+class SpenderWindow extends Analyzer {
+  static dependencies = {
+    enemies: Enemies,
+  };
+
+  enabled: boolean;
+
+  protected enemies!: Enemies;
+
+  activeSpenderWindow: ActiveSpenderWindow | null = null;
+  spenderWindows: FinishedSpenderWindow[] = [];
+
+  private stMSSpender: Talent = TALENTS.EARTH_SHOCK_TALENT;
+
+  constructor(options: Options) {
+    super(options);
+
+    if (this.selectedCombatant.hasTalent(TALENTS.ELEMENTAL_BLAST_TALENT)) {
+      this.stMSSpender = TALENTS.ELEMENTAL_BLAST_TALENT;
+    }
+
+    this.enabled = this.selectedCombatant.hasTalent(TALENTS.SURGE_OF_POWER_TALENT);
```

Comment:
this should be `this.active`. if you use that, `addEventListener` is a no-op so you don't need the conditional (but there may be other initialization that you need to skip)

```suggestion
    this.active = this.selectedCombatant.hasTalent(TALENTS.SURGE_OF_POWER_TALENT);
```